beater/java_attacher/java_attacher.go

```diff
@@ -0,0 +1,155 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/apm-server/beater/config"
+)
+
+type JavaAttacher struct {
+	cfg      config.JavaAttacherConfig
+	logLevel string
+	logger   *logp.Logger
+	logfn    logfn
+}
+
+type logfn func(...interface{})
+
+func New(cfg config.JavaAttacherConfig, logLevel string) (JavaAttacher, error) {
+	if cfg.JavaBin == "" {
+		if jh := os.Getenv("JAVA_HOME"); jh != "" {
+			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+		} else {
+			bin, err := exec.LookPath("java")
+			if err != nil {
+				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+			}
+			cfg.JavaBin = bin
+		}
+	} else {
+		// Ensure we're using the correct separators for the system
+		// running apm-server
+		cfg.JavaBin = filepath.FromSlash(cfg.JavaBin)
+	}
+	logger := logp.NewLogger("java-attacher")
+	return JavaAttacher{
+		cfg:      cfg,
+		logLevel: logLevel,
+		logger:   logger,
+		logfn:    setLogfn(logger, logLevel),
+	}, nil
+}
+
+// javaAttacher is bundled by the server
+// TODO: Figure out the real path
+var javaAttacher = filepath.FromSlash("/bin/apm-agent-attach-cli-1.24.0-slim.jar")
+
+func (j JavaAttacher) Run(ctx context.Context) error {
+	cmd := j.build(ctx)
+	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	donec := make(chan struct{})
+	defer close(donec)
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		b := struct {
+			Message string `json:"message"`
+		}{}
+		for scanner.Scan() {
+			select {
+			case <-ctx.Done():
+				return
+			case <-donec:
+				return
+			default:
+			}
+			if err := json.Unmarshal(scanner.Bytes(), &b); err != nil {
+				j.logger.Errorf("error unmarshaling attacher logs: %v", err)
+				continue
+			}
+			j.logfn(b.Message)
+		}
+		if err := scanner.Err(); err != nil {
+			j.logger.Errorf("error scanning attacher logs: %v", err)
+		}
+	}()
+
+	return cmd.Wait()
+}
+
+func (j JavaAttacher) build(ctx context.Context) *exec.Cmd {
+	args := append([]string{"-jar", javaAttacher}, j.formatArgs()...)
+	return exec.CommandContext(ctx, j.cfg.JavaBin, args...)
+}
+
+func (j JavaAttacher) formatArgs() []string {
+	args := []string{"--continuous", "--log-level", j.logLevel}
+
+	for _, flag := range j.cfg.DiscoveryRules {
+		for name, value := range flag {
+			args = append(args, makeArg("--"+name, value))
+		}
+	}
+	cfg := make([]string, 0, len(j.cfg.Config))
+	for k, v := range j.cfg.Config {
+		cfg = append(cfg, "--config "+k+"="+v)
+	}
+	// we want a predictable order for testing
+	sort.Strings(cfg)
+
+	return append(args, cfg...)
+}
+
+func makeArg(flagName string, args ...string) string {
+	return flagName + " " + strings.Join(args, " ")
+}
+
+func setLogfn(l *logp.Logger, level string) logfn {
+	switch level {
```

Comment:
The attacher emits the `log.level` in upper-case
```suggestion
	switch strings.ToLower(level) {
```

Reply:
This value is coming from the (lowercase) settings in apm-server.yml, I'm using that value for the `--log-level` flag when starting the attacher

Reply:
We should be varying the log function based on the log level emitted by the attacher though. We might pass `--log-level=debug` to the attacher, but that doesn't mean every log entry is debug level.

Reply:
In fact, we could always pass `--log-level=debug` and then let the server's logger decide whether or not to log the actual level. In theory, it could slow down the attacher a bit by logging and parsing the logs of fine-grained levels unnecessarily. But I don't expect that to be a problem in practice. It may make things a bit easier. Just an idea :)

Reply:
How about we go with `--log-level=debug` for now, and log the lines accordingly as they come. when we actually hook it up we can see if this is an issue or not?